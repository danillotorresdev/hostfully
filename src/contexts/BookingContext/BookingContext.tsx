import { ReactNode, createContext, useContext, useState } from "react";
import { z } from "zod";

const PassengerSchema = z.object({
  adults: z.number(),
  children: z.number(),
});

const BookingSchema = z.object({
  id: z.string(),
  startDate: z.string(),
  endDate: z.string(),
  origin: z.string(),
  destination: z.string(),
  passengers: PassengerSchema,
});

type Booking = z.infer<typeof BookingSchema>;

const BookingContext = createContext<{
  bookings: Booking[];
  createBooking: (booking: Booking) => void;
  deleteBooking: (bookingId: string) => void;
  validateBookingCreation: (booking: Booking) => boolean;
  validateBookingUpdate: (booking: Booking) => boolean;
  updateBooking: (booking: Booking) => void;
}>({
  bookings: [],
  createBooking: () => {},
  deleteBooking: () => {},
  validateBookingCreation: () => false,
  validateBookingUpdate: () => false,
  updateBooking: () => {},
});

export const useBookingContext = () => useContext(BookingContext);

type BookingProviderProps = {
  children: ReactNode;
};

export const BookingProvider = ({ children }: BookingProviderProps) => {
  const [bookings, setBookings] = useState<Booking[]>([
    {
      id: "1",
      origin: "Cairo",
      destination: "Great Pyramid of Giza",
      startDate: "2024-01-11T03:00:00.000Z",
      endDate: "2024-01-15T03:00:00.000Z",
      passengers: {
        adults: 1,
        children: 0,
      },
    },
    {
      id: "2",
      origin: "Bodrum",
      destination: "Mausoleum at Halicarnassus",
      startDate: "2024-01-11T03:00:00.000Z",
      endDate: "2024-01-15T03:00:00.000Z",
      passengers: {
        adults: 2,
        children: 1,
      },
    },
  ]);

  const createBooking = (booking: Booking) => {
    const validBookingData = BookingSchema.parse(booking);
    setBookings([validBookingData, ...bookings]);
  };

  const deleteBooking = (bookingId: string) => {
    setBookings(bookings.filter((booking) => booking.id !== bookingId));
  };

  const updateBooking = (booking: Booking) => {
    const validBookingData = BookingSchema.parse(booking);

    const foundBooking = bookings.find(
      (booking) => booking.id === validBookingData.id,
    );

    if (!foundBooking) {
      return;
    }

    const updatedBookings = bookings.map((booking) =>
      booking.id === validBookingData.id ? validBookingData : booking,
    );

    setBookings(updatedBookings);
  };

  const validateBookingCreation = (booking: Booking) => {
    const newBookingStartDate = new Date(booking.startDate);
    const newBookingEndDate = new Date(booking.endDate);

    for (const existingBooking of bookings) {
      const existingBookingStartDate = new Date(existingBooking.startDate);
      const existingBookingEndDate = new Date(existingBooking.endDate);
      if (
        (newBookingStartDate >= existingBookingStartDate &&
          newBookingStartDate <= existingBookingEndDate) ||
        (newBookingEndDate >= existingBookingStartDate &&
          newBookingEndDate <= existingBookingEndDate)
      ) {
        return false;
      }
    }

    return true;
  };

  const validateBookingUpdate = (booking: Booking) => {
    const newBookingStartDate = new Date(booking.startDate);
    const newBookingEndDate = new Date(booking.endDate);

    const selectedBooking = bookings.find(
      (existingBooking) => existingBooking.id === booking.id,
    );

    if (!selectedBooking) {
      return false;
    }

    const existingBookingStartDate = new Date(selectedBooking.startDate);
    const existingBookingEndDate = new Date(selectedBooking.endDate);

    if (
      newBookingStartDate.getTime() === existingBookingStartDate.getTime() &&
      newBookingEndDate.getTime() === existingBookingEndDate.getTime()
    ) {
      return true;
    }

    for (const existingBooking of bookings) {
      const existingBookingStartDate = new Date(existingBooking.startDate);
      const existingBookingEndDate = new Date(existingBooking.endDate);
      if (
        (newBookingStartDate >= existingBookingStartDate &&
          newBookingStartDate <= existingBookingEndDate) ||
        (newBookingEndDate >= existingBookingStartDate &&
          newBookingEndDate <= existingBookingEndDate)
      ) {
        return false;
      }
    }

    return true;
  };

  return (
    <BookingContext.Provider
      value={{
        bookings,
        createBooking,
        deleteBooking,
        updateBooking,
        validateBookingCreation,
        validateBookingUpdate,
      }}
    >
      {children}
    </BookingContext.Provider>
  );
};
