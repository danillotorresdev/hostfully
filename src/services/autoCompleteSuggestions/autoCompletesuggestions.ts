import { nominatimBaseAPI } from "@/services/nominatimBaseAPI";

type Place = {
  place_id: number;
  licence: string;
  osm_type: string;
  osm_id: number;
  lat: string;
  lon: string;
  category: string;
  type: string;
  place_rank: 10;
  importance: number;
  addresstype: string;
  name: string;
  display_name: string;
  boundingbox: string;
};

type Places = Place[];

export const placesAutoCompleteSuggestions = (search: string) =>
  nominatimBaseAPI.get<Places>("/search.php", {
    params: {
      city: search,
      format: "jsonv2",
    },
  });
