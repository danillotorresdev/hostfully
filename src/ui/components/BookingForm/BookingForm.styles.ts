import { styled } from "@linaria/react";

export const Form = styled.form`
  display: flex;
  justify-content: space-between;
  gap: 20px;
  flex-direction: column;
  & > * {
    flex: 1;
  }

  @media (min-width: 1024px) {
    flex-direction: row;
  }
`;

export const ButtonWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: end;
  & button {
    width: 100%;
    display: block;
    display: flex;
    justify-content: center;
    align-items: center;
    .svg {
      margin-left: 10px;
    }
  }
`;

export const InputWrapper = styled.div`
  position: relative;
  padding-bottom: 10px;
  @media (min-width: 1024px) {
    padding-bottom: 0;
  }
`;

export const ErrorMessage = styled.div`
  color: red;
  font-size: 0.8rem;
  position: absolute;
  bottom: -22px;
  @media (min-width: 1024px) {
    bottom: -33px;
  }
`;
