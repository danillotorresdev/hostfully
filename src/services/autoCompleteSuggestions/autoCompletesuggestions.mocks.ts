import { Places } from "@/services/autoCompleteSuggestions/autoCompletesuggestions";

export const placesMock: Places = [
  {
    addresstype: "city",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "city",
  },
  {
    addresstype: "town",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "town",
  },
  {
    addresstype: "village",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "village",
  },
  {
    addresstype: "municipality",
    boundingbox: "48.8155750,48.9021560,2.2241990,2.4699210",
    category: "tourism",
    display_name: "Paris, Île-de-France, France",
    importance: 0.935,
    lat: "48.8566969",
    licence:
      "Data © OpenStreetMap contributors, ODbL 1.0. https://osm.org/copyright",
    lon: "2.3514616",
    name: "Paris",
    osm_id: 7444,
    osm_type: "relation",
    place_id: 23424846,
    place_rank: 10,
    type: "municipality",
  },
];
