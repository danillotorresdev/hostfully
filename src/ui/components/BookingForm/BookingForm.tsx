import { useBookingContext } from "../../../contexts/BookingContext/BookingContext";
import * as S from "./BookingForm.styles";
import { z } from "zod";
import { useForm, SubmitHandler } from "react-hook-form";
import { DateRangePicker } from "@/ui/components/DateRangePIcker/DateRangePicker";
import { Button } from "@/ui/components/Button/Button";
import { AutoCompleteInput } from "@/ui/components/AutoCompleteInput/AutoCompleteInput";
import { PassengerCounterInput } from "@/ui/components/PassengersCounterWrapper/PassengersCounterWrapper";
import { useMemo } from "react";
import { v4 as uuid } from "uuid";
import { Toast } from "@/ui/components/Toast/Toast";
import { useToast } from "@/utils/hooks/useToast/useToast";

const createPassengerFormSchema = z.object({
  adults: z.number(),
  children: z.number(),
});

const createBookingFormSchema = z.object({
  destination: z.string(),
  origin: z.string(),
  startDate: z.string(),
  endDate: z.string(),
  passengers: createPassengerFormSchema,
});

type BookingFormProps = {
  bookingId?: string;
  afterSumbit?: () => void;
};

export function BookingForm({ bookingId, afterSumbit }: BookingFormProps) {
  const { showToast, toasts } = useToast();

  const {
    createBooking,
    validateBookingCreation,
    validateBookingUpdate,
    bookings,
    updateBooking,
  } = useBookingContext();

  const bookingToUpdate = useMemo(
    () =>
      bookingId ? bookings.find((booking) => booking.id === bookingId) : null,
    [bookingId, bookings],
  );

  const {
    register,
    handleSubmit,
    setValue,
    formState: { errors },
  } = useForm({
    defaultValues: {
      origin: bookingToUpdate?.origin ?? "",
      destination: bookingToUpdate?.destination ?? "",
      startDate: bookingToUpdate?.startDate ?? "",
      endDate: bookingToUpdate?.endDate ?? "",
      passengers: bookingToUpdate?.passengers ?? { adults: 1, children: 0 },
    },
  });

  register("startDate", {
    required: "Start date is required",
  });

  register("endDate", {
    required: "End date is required",
  });

  register("origin", {
    required: "Origin is required",
  });

  register("destination", {
    required: "Destination is required",
  });

  const onSubmit: SubmitHandler<z.infer<typeof createBookingFormSchema>> = (
    bookingProps,
  ) => {
    const newBooking = {
      id: bookingId ?? uuid(),
      ...bookingProps,
    };

    if (bookingId) {
      if (validateBookingUpdate(newBooking)) {
        updateBooking(newBooking);
        showToast({
          id: newBooking.id,
          message: "Booking updated",
          type: "success",
        });
      } else {
        showToast({
          id: newBooking.id,
          message: "Booking conflicts with existing bookings",
          type: "error",
        });
      }
    } else {
      if (validateBookingCreation(newBooking)) {
        createBooking(newBooking);
        showToast({
          id: newBooking.id,
          message: "Booking created",
          type: "success",
        });
      } else {
        showToast({
          id: newBooking.id,
          message: "Booking conflicts with existing bookings",
          type: "error",
        });
      }
    }

    if (afterSumbit) {
      afterSumbit();
    }
  };

  return (
    <>
      <S.Form className="form" onSubmit={handleSubmit(onSubmit)}>
        <S.InputWrapper>
          <AutoCompleteInput
            type="text"
            placeholder="City's name"
            label="Origin"
            defaultValue={bookingToUpdate?.origin}
            onPlaceChange={(place) => {
              setValue("origin", place);
            }}
          />
          <S.ErrorMessage>
            {errors.origin && <p>{errors.origin.message}</p>}
          </S.ErrorMessage>
        </S.InputWrapper>
        <S.InputWrapper>
          <AutoCompleteInput
            type="text"
            placeholder="City's name"
            defaultValue={bookingToUpdate?.destination}
            label="Destination"
            onPlaceChange={(place) => {
              setValue("destination", place);
            }}
          />
          <S.ErrorMessage>
            {errors.destination && <p>{errors.destination.message}</p>}
          </S.ErrorMessage>
        </S.InputWrapper>
        <S.InputWrapper>
          <DateRangePicker
            defaultValue={
              bookingToUpdate
                ? {
                    startDate: new Date(bookingToUpdate.startDate),
                    endDate: new Date(bookingToUpdate.endDate),
                  }
                : undefined
            }
            onSelect={(startDate, endDate) => {
              if (startDate && endDate) {
                setValue("startDate", startDate?.toISOString());
                setValue("endDate", endDate?.toISOString());
              }
            }}
          />
          <S.ErrorMessage>
            {errors.startDate && errors.endDate && (
              <p>Start and end date are required</p>
            )}
          </S.ErrorMessage>
        </S.InputWrapper>

        <PassengerCounterInput
          defaultValue={bookingToUpdate?.passengers}
          onChange={(count) => {
            setValue("passengers", count);
          }}
        />
        <S.ButtonWrapper>
          <Button type="submit" size="lg">
            {bookingId ? "Update" : "Create"} Booking
          </Button>
        </S.ButtonWrapper>
      </S.Form>
      {toasts.map((toast) => (
        <Toast key={uuid()} message={toast.message} type={toast.type} />
      ))}
    </>
  );
}
