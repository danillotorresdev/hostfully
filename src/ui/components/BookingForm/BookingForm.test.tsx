import { BookingForm } from "./BookingForm";
import { expect, it, vi } from "vitest";
import { beforeEach, describe } from "node:test";
import * as autoCompletesuggestionsAPI from "@/services/autoCompleteSuggestions/autoCompletesuggestions";
import { placesMock } from "@/services/autoCompleteSuggestions/autoCompletesuggestions.mocks";
import { AxiosRequestHeaders } from "axios";
import { BookingProvider } from "@/contexts/BookingContext/BookingContext";

import {
  render,
  waitFor,
  screen,
  act,
  fireEvent,
} from "@/utils/settings/tests/utilities";

vi.spyOn(
  autoCompletesuggestionsAPI,
  "placesAutoCompleteSuggestions",
).mockResolvedValue({
  data: placesMock,
  config: {
    headers: {} as AxiosRequestHeaders,
  },
  headers: {},
  status: 200,
  statusText: "OK",
  request: {},
});

vi.mock("@/utils/hooks/useToast/useToast", () => ({
  useToast: () => ({
    showToast: vi.fn(),
    toasts: [],
  }),
}));

const today = new Date("2030-05-29T12:10:40.070Z").toISOString();

describe("BookingForm", () => {
  beforeEach(() => {
    vi.spyOn(global, "Date").mockImplementation(
      () => today as unknown as string,
    );
  });

  describe("Create Booking", () => {
    it("should submit the form", async () => {
      vi.useFakeTimers();
      vi.setSystemTime(new Date("2024-09-14"));

      const afterSumbit = vi.fn();

      const { getByLabelText, getByText, debug } = render(
        <BookingForm afterSumbit={afterSumbit} />,
      );

      fireEvent.change(getByLabelText("Origin"), {
        target: { value: "Paris" },
      });
      fireEvent.change(getByLabelText("Destination"), {
        target: { value: "London" },
      });

      const departureLabel = screen.getByLabelText("date");
      fireEvent.click(departureLabel);

      expect(screen.getByText("Select")).toBeDefined();

      const calendar = screen.getByTestId("leftCalendar");
      const getStartDate = screen.getByText("15");
      const endDate = screen.getByText("20");

      fireEvent.click(getStartDate);
      fireEvent.click(endDate);

      const selectButon = calendar.querySelector("button") as HTMLElement;

      fireEvent.click(selectButon);

      act(() => {
        fireEvent.click(getByText("Select"));
      });

      const updatedStartDateLabel = screen.getByText("9/15/2024");
      const updatedEndDateLabel = screen.getByText("9/20/2024");

      expect(updatedStartDateLabel).toBeDefined();
      expect(updatedEndDateLabel).toBeDefined();

      const submitFormButton = screen.getByText("Create Booking");

      fireEvent.click(submitFormButton);

      waitFor(() => {
        expect(getByText("Booking created")).toBeDefined();
      });
    });
  });

  describe("Update Booking", () => {
    it("should submit the form", async () => {
      vi.useFakeTimers();
      vi.setSystemTime(new Date("2024-09-14"));

      const afterSumbit = vi.fn();

      const { getByLabelText, getByText } = render(
        <BookingProvider>
          <BookingForm bookingId="1" afterSumbit={afterSumbit} />
        </BookingProvider>,
      );

      fireEvent.change(getByLabelText("Origin"), {
        target: { value: "Paris" },
      });
      fireEvent.change(getByLabelText("Destination"), {
        target: { value: "London" },
      });

      const departureLabel = screen.getByLabelText("date");
      fireEvent.click(departureLabel);

      expect(screen.getByText("Select")).toBeDefined();

      const calendar = screen.getByTestId("leftCalendar");
      const getStartDate = screen.getByText("15");
      const endDate = screen.getByText("20");

      fireEvent.click(getStartDate);
      fireEvent.click(endDate);

      const selectButon = calendar.querySelector("button") as HTMLElement;

      fireEvent.click(selectButon);

      act(() => {
        fireEvent.click(getByText("Select"));
      });

      const updatedStartDateLabel = screen.getByText("1/11/2024");
      const updatedEndDateLabel = screen.getByText("1/15/2024");

      expect(updatedStartDateLabel).toBeDefined();
      expect(updatedEndDateLabel).toBeDefined();

      const submitFormButton = screen.getByText("Update Booking");

      fireEvent.click(submitFormButton);

      waitFor(() => {
        expect(getByText("Booking updated")).toBeDefined();
      });
    });
  });
});
