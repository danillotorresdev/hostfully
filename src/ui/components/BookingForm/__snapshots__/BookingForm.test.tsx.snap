// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="form f1q86ce6"
  >
    <div
      class="mocked-styled-2 i92tumy"
    >
      <div>
        <div
          class="mocked-styled-21 i19zws4p"
        >
          <label
            class="mocked-styled-22 l5v2eyb"
            for="Origin"
          >
            Origin
          </label>
          <input
            class="mocked-styled-23 i1hv9n31"
            id="Origin"
            placeholder="City's name"
            type="text"
            value=""
          />
        </div>
        <div
          class="mocked-styled-24 skesand"
        >
          <ul
            class="mocked-styled-25 s1qvsmxh"
          />
        </div>
      </div>
      <div
        class="mocked-styled-3 e1hmdc16"
      />
    </div>
    <div
      class="mocked-styled-2 i92tumy"
    >
      <div>
        <div
          class="mocked-styled-21 i19zws4p"
        >
          <label
            class="mocked-styled-22 l5v2eyb"
            for="Destination"
          >
            Destination
          </label>
          <input
            class="mocked-styled-23 i1hv9n31"
            id="Destination"
            placeholder="City's name"
            type="text"
            value=""
          />
        </div>
        <div
          class="mocked-styled-24 skesand"
        >
          <ul
            class="mocked-styled-25 s1qvsmxh"
          />
        </div>
      </div>
      <div
        class="mocked-styled-3 e1hmdc16"
      />
    </div>
    <div
      class="mocked-styled-2 i92tumy"
    >
      <div
        class="mocked-styled-4 cx3igb0"
      >
        <label
          class="mocked-styled-6 d1m4j7uz"
        >
          Date
        </label>
        <div
          aria-label="date"
          class="mocked-styled-5 d6fdk0w"
        >
          <div
            class="mocked-styled-9 dv2dzrz"
          >
            <div
              class="mocked-styled-10 d1xj2x14"
            >
              <svg
                class="svg svg-calendar"
                data-testid="icon-svg"
                style="width: 24px; height: 24px;"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#calendar"
                />
              </svg>
              Check-in
            </div>
            <div
              class="mocked-styled-10 d1xj2x14"
            >
              <svg
                class="svg svg-calendar"
                data-testid="icon-svg"
                style="width: 24px; height: 24px;"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#calendar"
                />
              </svg>
              Check-out
            </div>
          </div>
        </div>
      </div>
      <div
        class="mocked-styled-3 e1hmdc16"
      />
    </div>
    <div
      class="mocked-styled-27 c1u669jx"
    >
      <label
        class="mocked-styled-26 lzhktur"
      >
        Passengers
      </label>
      <button
        class="mocked-styled-29 pfu8a7v"
        type="button"
      >
        <svg
          class="svg svg-member"
          data-testid="icon-svg"
          style="width: 24px; height: 24px;"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="/assets/sprite.svg#member"
          />
        </svg>
        1 passenger
      </button>
    </div>
    <div
      class="mocked-styled-1 bezp364"
    >
      <button
        class="mocked-styled-20 bb8cb6j"
        style="--bb8cb6j-0: 55px; --bb8cb6j-1: 16px; --bb8cb6j-2: 15px 30px; --bb8cb6j-3: #2D2AA5; --bb8cb6j-4: #fff; --bb8cb6j-5: #40caa1;"
        type="submit"
      >
        Create Booking
      </button>
    </div>
  </form>
</DocumentFragment>
`;
